onnxruntime/core/graph/function.cc

```diff
@@ -399,11 +593,34 @@ FunctionImpl::FunctionImpl(const onnxruntime::Graph& graph,
           graph_inputs[iter->second] = &unused_formal_param;
         }
       } else {
-        auto& n_input = function_body_graph.GetOrCreateNodeArg(
-            tensor_name + "_" + std::to_string(node_index), nullptr);
-        inputs.push_back(&n_input);
+        // // If input is part of function outputs, preserve NodeArg and input/output names
+        iter = output_name_idx_map.find(tensor_name);
+        if (iter != output_name_idx_map.end()) {
+          const std::string& actual_parameter_name = function_op_node_proto.output().Get(iter->second);
+          const onnxruntime::NodeArg* node_arg = parent_graph_->GetNodeArg(actual_parameter_name);
+          const ONNX_NAMESPACE::TypeProto* actual_type = node_arg->TypeAsProto();
+          auto& n_input = function_body_graph.GetOrCreateNodeArg(actual_parameter_name, actual_type);
+          inputs.push_back(&n_input);
+        } else {
+          // Input is intermidiate input in function body.
```

Comment:
>intermidiate [](http://example.com/codeflow?start=22&length=12)

nit: intermidiate -> intermediate

[https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2017andLater](https://marketplace.visualstudio.com/items?itemName=EWoodruff.VisualStudioSpellCheckerVS2017andLater) saves me most of the time.